go-types/filesource/filesource.go

```diff
@@ -0,0 +1,563 @@
+		return err
+	}
+
+	r.log("completed sweep of %q through modification time %s", r.prefix, r.state.MinMod)
+	return nil
+}
+
+func (r *reader) log(msg string, args ...interface{}) {
+	r.shared.mu.Lock()
+	defer r.shared.mu.Unlock()
+	_ = r.shared.enc.Encode(airbyte.NewLogMessage(airbyte.LogLevelInfo, msg, args...))
+}
+
+func (r *reader) shouldSkip(obj ObjectInfo) (_ bool, reason string) {
+	// Have we already processed through this filename in this sweep?
+	if r.state.Path > obj.Path {
+		return true, "state.Path > obj.Path"
+	}
+	if r.state.Path == obj.Path && r.state.Complete {
+		return true, "state.Path == obj.Path && Complete"
+	}
+	// Is the path excluded by our regex ?
+	if r.pathRe != nil && !r.pathRe.MatchString(obj.Path) {
+		return true, "regex not matched"
+	}
+	// Is the path modified before our window (inclusive) ?
+	if r.state.MinMod.After(obj.ModTime) {
+		return true, "MinMod.After(obj.ModTime)"
+	}
+	// Is the path modified after our window (exclusive) ?
+	if r.state.MaxMod.Before(obj.ModTime) {
+		return true, "MaxMod.After(obj.ModTime)"
+	}
+
+	return false, ""
+}
+
+func (r *reader) processObject(ctx context.Context, obj ObjectInfo) error {
+	rr, obj, err := r.store.Read(ctx, obj)
+	if err != nil {
+		return err
+	}
+	defer rr.Close()
+
+	r.log("processing file %q", obj.Path)
+
+	tmp, err := ioutil.TempFile("", "parser-config-*.json")
+	if err != nil {
+		return fmt.Errorf("creating parser config: %w", err)
+	}
+	defer os.Remove(tmp.Name())
+
+	if err = r.makeParseConfig(obj).WriteToFile(tmp); err != nil {
+		return fmt.Errorf("writing parser config: %w", err)
+	}
+
+	// Skip tracks records remaining to skip, because they've already
+	// been ingested in a prior invocation of this capture.
+	var skip int
+
+	if r.state.Path == obj.Path {
+		if r.state.Complete {
+			panic("should have been filtered already")
+		}
+		skip = r.state.Records
+	}
+
+	r.state.Path = obj.Path
+	r.state.Records = 0
+	r.state.Complete = false
+
+	err = parser.ParseStream(ctx, tmp.Name(), rr, func(lines []json.RawMessage) error {
+		r.state.Records += len(lines)
+
+		if ll := len(lines); ll < skip {
+			skip -= ll
+			return nil
+		} else if skip != 0 {
+			lines = lines[skip:]
+			skip = 0
+		}
+
+		return r.commitLines(lines)
+	})
+	if err != nil {
+		return err
+	}
+
+	// Write a final checkpoint to mark the completion of the file.
+	r.state.Complete = true
+	r.state.Records = 0
+
+	if err := r.commitLines(nil); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *reader) makeParseConfig(obj ObjectInfo) *parser.Config {
+	var cfg = new(parser.Config)
+	if c := r.config.ParserConfig(); c != nil {
+		*cfg = c.Copy()
+	}
+
+	cfg.Filename = obj.Path
+	if obj.ContentType != "" {
+		cfg.ContentType = obj.ContentType
+	}
+	if obj.ContentEncoding != "" {
+		cfg.ContentEncoding = obj.ContentEncoding
+	}
+	// If the user supplied a location for this, then we'll use that. Otherwise, use the default
+	if cfg.AddRecordOffset == "" {
```

Comment:
I'm realizing that if the user supplies a different location for this, then the document schema and SourceDefinedPrimaryKey will no longer be correct. I don't know if that really matters, though, since the user can change the schema, and I don't know if we really do anything with SourceDefinedPrimaryKey anyway.

Reply:
We do use SourceDefinedPrimaryKey -- it's plumbed through to the `key` of the collection.

Yea, it will break currently. It would require dynamically parsing the pointer and adjusting the schema. Right now I'd rather just say "you can't change this at all".